engine-tests/src/test/java/org/terasology/engine/rendering/world/ChunkMeshWorkerTest.java

```diff
@@ -23,107 +29,158 @@
+        return Mono.just(Tuples.of(chunk, mock(ChunkMesh.class)));
     }
 
     @BeforeEach
     void makeWorker() {
-        worker = new ChunkMeshWorker(ChunkMeshWorkerTest::alwaysCreateMesh, null);
+        comparator = Comparator.comparingDouble(chunk ->
+                chunk.getRenderPosition().distanceSquared(currentPosition.x, currentPosition.y, currentPosition.z)
+        );
+
+        StepVerifier.setDefaultTimeout(Duration.ofSeconds(10));
+
+        // Use virtual time so we don't have to wait around in real time
+        // to see whether there are more events pending.
+        // Requires that the schedulers be created _inside_ the withVirtualTime supplier.
+        verifier = StepVerifier.withVirtualTime(() -> {
+            worker = new ChunkMeshWorker(
+                    ChunkMeshWorkerTest::alwaysCreateMesh,
+                    comparator,
+                    Schedulers.parallel(),
+                    Schedulers.single()
+            );
+            return worker.getCompletedChunks();
+        });
     }
 
-    @Test
-    void testChunkIsNotProcessedTwice() {
-        var chunk1 = new DummyChunk(position0);
-        var chunk2 = new DummyChunk(position0);
-        worker.add(chunk1);
-        worker.add(chunk2);
-        // drain
-        fail("TODO: assert number of results == 1");
+    static Chunk newDirtyChunk(Vector3ic position) {
+        var chunk = new DummyChunk(position);
+        chunk.markReady();
+        chunk.setDirty(true);
+        return chunk;
     }
 
     @Test
-    void testChunkIsRegeneratedIfDirty() {
-        var chunk = new DummyChunk(position0);
-        worker.add(chunk);
-        // tick - work function has started
-        chunk.setDirty(true);
-        worker.add(chunk);
-        // drain
-        fail("TODO: assert number of results == 2");
+    void testMultipleChunks() {
+        var chunk1 = newDirtyChunk(position0);
+        var chunk2 = newDirtyChunk(new Vector3i(position0).add(1, 0, 0));
+
+        verifier.then(() -> {
+                    worker.add(chunk1);
+                    worker.add(chunk2);
+                    worker.update();
+                })
+                .expectNext(chunk1, chunk2)
+                .verifyTimeout(Duration.ofSeconds(5));
     }
 
     @Test
-    void testChunksDirtiedBetweenGenerationAndUploadAreUploadedAnyway() {
-        // maybe redundant with testChunkIsRegeneratedIfDirty?
-        // I guess the assertion is that the upload function did a thing
-        // instead of skipping a thing.
+    void testChunkIsNotProcessedTwice() {
+        var chunk1 = newDirtyChunk(position0);
+
+        verifier.then(() -> {
+                    worker.add(chunk1);
+                    worker.add(chunk1);  // added twice
+                    worker.update();
+                })
+                .expectNextCount(1).as("expect only one result")
+                .then(() -> {
+                    // adding it again and doing another update should still not change
```

Comment:
do we need another assertion after this? :thinking:

Reply:
The `verify` at the end will fail if another thing comes through that's not expected, with a message like

> expected: timeout(5s);
> actual: onNext(DummyChunk{chunkPos=( 1.240E+2  4.560E+2  7.890E+2), dirty=false, mesh=Mock for ChunkMesh, hashCode: 1832543155, ready=true})